Added Feature:  Note Type Creator

Squeenix opened the pull request:
I've implemented most of the features applicable to mobile from the note type editor available on desktop Anki (Tools -> Manage Note Types).  It is present in the main nav drawer, but it might be better somewhere else.

There is one major feature missing, which is the ability to edit templates/cards.  This was not implemented by myself because there is an almost-complete template editor existing.  That feature should definitely be completed before this is released.  There are a few options which did not appear to be applicable to mobile didn't add.  Other areas of potential concern I can see are UI and localization.


Modifications to libanki are:

1.  Uncommenting setSortIdx and changing an outdated method call

2.  Removing a line where addField sets the field blank (I don't know why it does this, it doesn't in desktop Anki and nothing else calls addField)


I've personally tested it and couldn't find issues, but I wouldn't consider it 'extensive'.  I've never contributed to community projects or used Git before, but I think this should possible be placed in a feature branch(?)

timrae commented:
Hi,  wow thanks a lot for this! It's a very useful feature.
I haven't gone through the code itself in much detail yet, but here's some quick feedback on the design itself:

* Since it's a bit of an "advanced" feature, how about exposing it in the overflow menu in the DeckPicker instead of the NavigationDrawer to? I don't personally edit note types very often, but I'm not sure about others...? Also, how about renaming the text to "Manage note types" as in Anki Desktop?

* If you agree with the above, you should probably also change `ModelBrowser` to extend from `AnkiActivity` instead of `NavigationDrawerActivity`. Currently the Activity finishes when you try and open the Navigation Drawer from ModelBrowser actually.

* Attempting to delete a note type correctly shows a warning that a full sync will be required. However, if you accept and then go to delete a different note type, the same message appears. This is unexpected (different from desktop). Maybe you're not actually committing the changes to the database?

* In Android 5 the color of the status bar in these new activities is white. If I remember correctly I solved this problem by surrounding the layouts in a `CoordinatorLayout`. 

* The `ModelEditor` activity might be better called the `FieldEditorActivity`? It's not super clear that we're looking at all of the fields that belong to the model after tapping on a model, so maybe put "Edit fields" as the title. Also, tapping the + button here shows an input dialog with the title "Add note type". This should really be "Add field"

* Tapping on "Edit cards" takes us to the existing card editor, but this doesn't actually *really* let us edit cards at the moment. The code is all there actually, but when I made it I was afraid that without including the preview function that Anki Desktop has to preview how the code renders, the feature would be too hard to use and may lead to us getting inundated with support requests from people that accidentally broke a model without reading the Anki manual. Would you be interested in adding a preview feature and then exposing the functionality to edit the card templates?

* I noticed when skimming the code that you made a change to a libanki function. What's the necessity / rationale behind that change?

Squeenix commented:
Thanks for the input

> Since it's a bit of an "advanced" feature, how about exposing it in the overflow menu...

Thinking about it, it's probably was a bad idea to leave it exposed in the nav menu considering you can wreck your models with it.  I've moved it to the DeckPicker overflow menu and fixed the title.


> Attempting to delete a note type correctly shows a warning that a full sync will be required...

They were being committed to DB, it was a UI issue.  Fixed in latest commit.


>  In Android 5 the color of the status bar in these new activities is white...

Fixed


>  The ModelEditor activity might be better called the FieldEditorActivity...

I've changed it to ModelFieldEditor instead to indicate the pairing with ModelBrowser and to prevent confusion with FieldEditText


>  Tapping the + button here shows an input dialog with the title "Add note type". This should really be "Add field"

Fixed


> Tapping on "Edit cards" takes us to the existing card editor, but this doesn't actually really let us edit cards at the moment...

I'd be fine with doing the preview function.  Is the UI functionality for creating and saving new cards/templates currently implemented?


> inundated with support requests from people that accidentally broke a model without reading the Anki manual

Might it be useful to include a link to the Anki manual like in the desktop version?  I imagine that that was probably implemented because of people breaking their models and asking for support

hssm commented:
Some usability comments:
- Open the dialog with a single tap in the field editor instead of long-tap.
- In the reposition dialog, you should change the keyboard type to only show a numeric keypad to make things easier.
- *Maybe* we need a confirmation dialog for deleting a note type.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+    private ArrayList<JSONObject> models;
+    private ArrayList<Integer> cardCounts;
+    private ArrayList<Long> modelIds;
+    private ArrayList<DisplayPair> modelDisplay;
```
Try to follow the same [naming convention](https://source.android.com/source/code-style.html#follow-field-naming-conventions) that we use in the rest of the application.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+                String cardName = modelDisplay.get(position).getName();
+                currentID = modelIds.get(position);
+                currentPos = position;
+                cMenu = new ModelBrowserContextMenu().newInstance(cardName, mContextMenuListener);
```
This is a static method so you don't need to instantiate an object. Just do `ModelBrowserContextMenu.newInstance()`.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+        modelAddName.add(addForwardOptionalReverseName);
+        modelAddName.add(addClozeModelName);
+
+        for (int i = 0; i < models.size(); i++) {
```
You can use the for-each style loop for iterable collections when you don't need the index:
`for (JSONObject model : models)`

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+                        modelNameInput.setSingleLine();
+
+                        //Temporary workaround - Lack of stdmodels class
+                        if (addSelectionSpinner.getSelectedItemPosition() < 4) {
```
You need a comment explaining what this 4 does. It's not immediately obvious to someone reading it for the first time.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+                                }
+                                refreshList();
+                            }
+                            else{
```
I've noticed a lot of small formatting errors like this. Instead of nitpicking over each one, I think letting your IDE auto-format the classes you've created should cover most of them.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+    }
+
+    private void closeActivity(int reason) {
+        DeckTask.cancelTask();
```
Remove this line. I am pretty sure this will cause problems. (In fact, this method shouldn't even be public).

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,724 @@
+     * Used so that the main ListView is able to display the number of notes using the model
+     * along with the name.
+     */
+    public class DisplayPair {
```
There is a built-in [Pair](http://developer.android.com/reference/android/util/Pair.html) class which does something similar to this. You *might* save a few lines of code by using it. I'll let you decide if it's worth it.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java:
```diff
@@ -0,0 +1,587 @@
+
+public class ModelFieldEditor extends AnkiActivity{
+
+    private final static String EXTRA_CALLER = "EXTRA_CALLER";
```
Is this needed?

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java:
```diff
@@ -0,0 +1,587 @@
+                    @Override
+                    public void onPositive(MaterialDialog dialog) {
+                        String fieldName = fieldNameInput.getText().toString()
+                                .replaceAll("[\'\"\\n\\r\\[\\]\\(\\)]", "");
```
Is there an equivalent regular expression on the desktop client that this is copying?

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java:
```diff
@@ -0,0 +1,587 @@
+
+    /**
+     * Displays loading bar when deleting a model loading bar is needed
+     * because deleting a model also deletes all of the associated cards/notes *
```
This comment is copy/paste shrapnel.

hssm commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ModelBrowserCloneInput.java:
```diff
@@ -0,0 +1,52 @@
+import com.ichi2.anki.R;
+import com.ichi2.themes.Themes;
+
+public class ModelBrowserCloneInput extends DialogFragment {
```
This class doesn't appear to be used anywhere.

Squeenix commented:
Thanks a lot for checking over these things.  I've never worked on anything over a couple thousand lines, so I appear to have made a lot of stylistic mistakes.

I'll try to improve the style/readability and address the comments.

hssm commented on AnkiDroid/src/main/java/com/ichi2/async/DeckTask.java:
```diff
@@ -1116,6 +1141,151 @@ private TaskData doInBackgroundRemoveTemplate(TaskData... params) {
+            try {
+                long modID = n.getLong("id");
+                cardCount.add(col.getModels().nids(col.getModels().get(modID)).size());
+            } catch (JSONException e) {
```
I find repeatedly catching JSONExceptions to be distracting. It should be easier to follow if you just encase the whole block of code with a single try/catch instead of repeating it twice (ultimately it achieves the same thing here).

hssm commented on AnkiDroid/src/main/java/com/ichi2/libanki/Models.java:
```diff
@@ -509,16 +509,17 @@ public int sortIdx(JSONObject m) {
-    // }
+    public void setSortIdx(JSONObject m, int idx) throws ConfirmModSchemaException{
+        try {
+            mCol.modSchema();
```
The desktop client does `mCol.modSchema(true)`. Unless there is a good reason not to (and we document it), we should do the same.

hssm commented on AnkiDroid/src/main/res/values/17-model-manager.xml:
```diff
@@ -0,0 +1,72 @@
+<resources>
+
+    <!--Menu Labels-->
+    <string name="model_browser_label">Manage Note Types</string>
```
I think you will need to create a new string (and put it in 02-strings.xml) to serve as the menu text because it needs to be lower case (`Manage note types`) to be consistent with the other menu items. It looks out of place at the moment.

hssm commented:
And one last thing. A couple files have had their file mode changed. I'm not sure why (probably wasn't intentional) but we shouldn't be committing those changes so try to revert those.

Other than that, it's all looking good. The libanki changes are fine too.

timrae commented:
> I'd be fine with doing the preview function. Is the UI functionality for creating and saving new cards/templates currently implemented?

Thanks! Probably best to do that in a separate PR.

The `EditText`s which hold the front / back / style code are all set to read only, but the methods are all there in the Activity for saving template edits if you expose them in the GUI. The easiest way to do preview is probably to make a new DeckTask which adds a new temporary note, gets the rendered card html, and then immediately deletes the note from the DB.

timrae commented:
@Squeenix 
I would like to move to beta (i.e. freeze new features) in the next week or two; will you have time to work to implement the changes requested by @hssm? Also, unfortunately it seems that that there are merge conflicts now.... Are you comfortable resolving these with git?

If no to either of the previous questions, I can probably help out...

timrae commented:
I noticed you pushed some new commits about a week ago... Is this ready for re-review now or are you still planning some changes? Please let me know when you're ready for review. It seems there are still some merge conflicts btw

Squeenix commented:
Ah, I was a bit busy lately but the pushed changes implement everything everything by hssm suggested except

> .replaceAll("[\'\"\\n\\r\\[\\]\\(\\)]", "");
>  Is there an equivalent regular expression on the desktop client that this is copying?

I need to go and look at the desktop client about this.  I took the expression that was being used to strip stuff from deck names from AnkiDroid (I think it's okay but I'll look the desktop version because models are permanent so it would be bad to have illegal characters).

Also the changed file modes (I have no idea what these are).

I just synced the fork so there shouldn't be any merge conflicts (I think)

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -0,0 +1,678 @@
+                                    model.put("name", deckName);
+                                    col.getModels().update(model);
+                                    mModels.get(mModelListPosition).put("name", deckName);
+                                    mModelDisplayList.set(mModelListPosition, new DisplayPair(mModels.get(mModelListPosition).getString("name"), mCardCounts.get(mModelListPosition)));
```
This line is way too long, please try to stick to the 120 character limit

timrae commented:
I kind of want to push forward with the beta release, so I've taken it upon myself to work through the remaining issues on this myself. I'll probably submit a separate PR this evening and merge it.

timrae commented:
BTW, the `replaceAll("[\'\"\\n\\r\\[\\]\\(\\)]", "")` regex is unnecessary... It appears that Anki Desktop permits arbitrary characters in the model name
